source/dofs/dof_handler.cc

```diff
@@ -2593,6 +2593,31 @@ DoFHandler<dim, spacedim>::distribute_mg_dofs()
 
 
 
+template <int dim, int spacedim>
+void
+DoFHandler<dim, spacedim>::distribute_non_local_dofs()
+{
+  Assert(
+    this->object_dof_indices.size() > 0,
+    ExcMessage(
+      "Distribute active DoFs using distribute_dofs() before calling distribute_non_local_dofs()."));
+
+  auto non_local_dh = begin()->get_fe().get_non_local_dof_handler();
+  for (auto cell : active_cell_iterators())
+    {
+      // Make sure all cells use the same NonLocalDoFHandler.
+      Assert(non_local_dh == cell->get_fe().get_non_local_dof_handler(),
+             ExcMessage(
+               "You are trying to use different NonLocalDoFHandler objects"));
```

Comment:
This is wanted? Only a single non-local dof handler is possible? What happens in hp case?

Reply:
For the moment, every element of the HP collection *must* return the same NonLocalDoFHandler. The reasoning behind this is that non local dofs are not associated to the finite element itself, but the tringulation as a whole (they may be a  number of global functions), so it is not allowed, at least at this stage, to have different NonLocalDoFHandler.